gen/wrap.jl

```diff
@@ -0,0 +1,187 @@
+    "gdal_utils.h",
+    "ogr_api.h",
+    "ogr_core.h",
+    "ogr_srs_api.h",
+])
+
+# skip these; right hand side is undefined usually because of Skipping MacroDefinition
+# or because the right hand side is a function that is defined later
+const skip_exprs = [
+    :(const CPLAssert = expr),
+    :(const EMULATED_BOOL = bool),
+    :(const GINT64_MIN = GINTBIG_MIN),
+    :(const GINT64_MAX = GINTBIG_MAX),
+    :(const GUINT64_MAX = GUINTBIG_MAX),
+    :(const VSI_L_OFFSET_MAX = GUINTBIG_MAX),
+    :(const CPL_STATIC_ASSERT_IF_AVAILABLE = x),
+    :(const CPL_NULLPTR = NULL),
+    :(const stat = Cvoid),
+    :(const tm = Cvoid),
+    :(const CPLReadDir = VSIReadDir),
+    :(const CPLFree = VSIFree),
+    :(const VALIDATE_POINTER_ERR = CE_Failure)
+]
+
+const replace_exprs = Dict{Expr,Expr}(
+    :(const VSIStatBuf = stat) => :(const VSIStatBuf = Cvoid),
+    :(const VSIStatBufL = __stat64) => :(const VSIStatBufL = Cvoid),
+)
+
+"""
+Custom rewriter for Clang.jl's C wrapper
+
+Gets called with all expressions in a header file, or all expressiong in a common file.
+If available, it adds docstrings before every expression, such that Clang.jl prints them
+on top of the expression. The expressions themselves get sent to `rewriter(::Expr)`` for
+further treatment.
+"""
+function rewriter(xs::Vector)
+    rewritten = Any[]
+    for x in xs
+        # Clang.jl inserts strings like "# Skipping MacroDefinition: X"
+        # keep these to get a sense of what we are missing
+        if x isa String
+            push!(rewritten, x)
+            continue
+        end
+        @assert x isa Expr
+
+        x in skip_exprs && continue
+        x = get(replace_exprs, x, x)
+
+        name = cname(x)
+        node = findnode(name, doc)
+        docstr = node === nothing ? "" : build_docstring(node)
+        isempty(docstr) || push!(rewritten, addquotes(docstr))
+        x2 = rewriter(x)
+        push!(rewritten, x2)
+    end
+    rewritten
+end
+
+"Rewrite expressions in the ways listed at the top of this file."
+function rewriter(x::Expr)
+    if @capture(x,
+        function f_(fargs__)
+            ccall(fname_, rettype_, argtypes_, argvalues__)
+        end
+    )
+        # it is a function wrapper around a ccall
+
+        # lowercase the function name
+        f′ = Symbol(lowercase(String(f)))
+
+        # bind the ccall such that we can easily wrap it
+        cc = :(ccall($fname, $rettype, $argtypes, $(argvalues...)))
+
+        # eval to check if it is Ptr{Cvoid} indirectly
+        # e.g. const GDALInfoOptions = Cvoid; Ptr{GDALInfoOptions}
+        # it works since we include common.jl, meaning we do need to run this twice
+        # in case a new run creates new const X = Cvoid
+        isptrvoid = try
+            eval(rettype) == Ptr{Cvoid}
+        catch  # UndefVarError
+            # it might still be true if common.jl is not yet complete, run twice to avoid this
+            false
+        end
+
+        cc′ = if rettype == :Cstring
```

Comment:
nit: can we do `cc2` instead of `cc'`? That way we don't run the potential interpretation of `cc'` as the unary operator `'` being applied to the variable `cc`.

Reply:
Yeah don't want it to be confusing. It was a prime though, not a quote.